File: [A]tree/[A]tree-bfs/199-binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        if not root:
            return res
        queue = deque([root])
        while queue:
            level_count = len(queue)
            for i in range(level_count):
                node = queue.popleft()
                if i == level_count - 1:
                    res.append(node.val)
                for child in [node.left, node.right]:
                    if child:
                        queue.append(child)
        return res
    
# time O(n), due to traverse
# space O(n), due to queue's size (tree diameter or last level)
# using tree and bfs

File: [A]tree/[A]tree-bfs/199-binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        
        ArrayDeque<TreeNode> queue = new ArrayDeque<>();

        if (root != null) {
            queue.offer(root);
        }

        while (!queue.isEmpty()) {
            int levelCount = queue.size();
            for (int i = 0; i < levelCount; i++) {
                TreeNode node = queue.poll();
                if (i == levelCount - 1) {
                    res.add(node.val);
                }
                for (TreeNode child: new TreeNode[]{node.left, node.right}) {
                    if (child != null) {
                        queue.offer(child);
                    }
                }
            }
        }
        return res;
    }
}

// time O(n), due to traverse
// space O(n)
// using tree and bfs